### PR TITLE
Amd64/misc kernel fixes

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

Misc kernel fixes

## Proposed changes

- [WIN32K] Add missing NULL checks
- [NTOS:KDBG] Improve x64 stack trace printing
- [HAL] Add missing I/O port to int 10 BIOS emulator
- [KERNEL32] Fix printing exception address
- [NTOS] Add a hack for VBox
- [NTOS:KE] Restore interrupts in KiSwitchKernelStack
- [NTOS:KE/X64] Enable interrupts when accessing user mode memory
- [NTOS] Treat page faults with interrupts disabled as access violation
- [NTOS] Don't assert, when dispatching an exception to user mode fails

### CODE REVIEW

File: hal/halx86/amd64/x86bios.c [hunk 1]
```diff
@@ -321,9 +321,15 @@ ValidatePort(
...
 
-        // CHECKME!
+        // OVMF debug messages used by VBox / QEMU
+        // https://www.virtualbox.org/svn/vbox/trunk/src/VBox/Devices/EFI/Firmware/OvmfPkg/README
+        case 0x402: return (Size == 1) && IsWrite;
+
+        // BOCHS VBE: https://forum.osdev.org/viewtopic.php?f=1&t=14639
         case 0x1CE: return (Size == 1) && IsWrite;
         case 0x1CF: return (Size == 1);
+
```
Comment (HBelusca) [inline hal/halx86/amd64/x86bios.c:331]: You don't sort the ports by increased number?

File: ntoskrnl/ke/amd64/except.c [hunk 2]
```diff
@@ -119,6 +119,9 @@ KiDispatchExceptionToUser(
     /* Get pointer to the usermode context, exception record and machine frame */
     UserStack = (PKUSER_EXCEPTION_STACK)UserRsp;
 
+    /* Enable interrupts */
+    _enable();
```
Comment (HBelusca) [inline ntoskrnl/ke/amd64/except.c:123]: In the `_SEH2_EXCEPT` / `_SEH2_YIELD(return);` case below, there is no need to re-disable the interrupts?

File: ntoskrnl/ke/amd64/stubs.c [hunk 3]
```diff
@@ -157,6 +159,9 @@ KiSwitchKernelStack(PVOID StackBase, PVOID StackLimit)
     /* Adjust Rsp0 in the TSS */
     Pcr->TssBase->Rsp0 += StackOffset;
 
+    /* Restore interrupts */
+    __writeeflags(Eflags);
```
Comment (HBelusca) [inline ntoskrnl/ke/amd64/stubs.c:163]: This also does an `_enable()` too?

File: win32ss/gdi/ntgdi/bitmaps.c [hunk 4]
```diff
@@ -333,6 +333,11 @@ IntCreateCompatibleBitmap(
                           Planes ? Planes : 1,
                           Bpp ? Bpp : dibs.dsBm.bmBitsPixel,
                           NULL);
+            if (Bmp == NULL)
+            {
+                DPRINT1("Failed to allocate a bitmap!\n");
+                return NULL;
+            }
```
Comment (HBelusca) [inline win32ss/gdi/ntgdi/bitmaps.c:340]: Interesting that in line 350 below, this is `psurf` that is acted upon (for setting `SurfObj.hdev`) and not `psurfBmp`. Is there an explanation for this?

File: ntoskrnl/kdbg/kdb_cli.c [hunk 5]
```diff
@@ -1211,45 +1211,31 @@ KdbpCmdBackTrace(
...
-        CurrentRsp = Context.Rsp;
-
-        BOOLEAN GotNextFrame = GetNextFrame(&Context);
+        BOOLEAN GotNextFrame;
 
-        KdbpPrint("[%p] ", (PVOID)CurrentRsp);
-        Context.Rsp = Context.Rsp;
+        KdbpPrint("[%p] ", (PVOID)Context.Rsp);
 
         /* Print the location afrer the call instruction */
```
Comment (HBelusca) [inline ntoskrnl/kdbg/kdb_cli.c:1223]:
```suggestion
        /* Print the location after the call instruction */
```

File: ntoskrnl/kdbg/kdb_cli.c [hunk 6]
```diff
@@ -1211,45 +1211,31 @@ KdbpCmdBackTrace(
...
         if (!GotNextFrame)
         {
-            KdbpPrint("Couldn't access memory at 0x%p!\n", (PVOID)Context.Rsp);
+            KdbpPrint("Couldn't get next frame\n");
             break;
         }
-    }
-
+    } while ((Context.Rip != 0) && (Context.Rsp != 0));
+
```
Comment (HBelusca) [inline ntoskrnl/kdbg/kdb_cli.c:1238]:
```suggestion

```

File: win32ss/gdi/ntgdi/dibobj.c [hunk 7]
```diff
@@ -69,6 +69,11 @@ CreateDIBPalette(
                                     0,
                                     0,
                                     0);
+        if (ppal == NULL)
+        {
+            DPRINT1("Failöed to allocate palette.\n");
```
Comment (SergeGautherie) [inline win32ss/gdi/ntgdi/dibobj.c:74]:
```suggestion
            DPRINT1("Failed to allocate palette.\n");
```

File: ntoskrnl/ke/amd64/stubs.c (hunk 3 above)
Comment (tkreuzer) [reply ntoskrnl/ke/amd64/stubs.c:163]: yes, if it was enabled before.

File: win32ss/gdi/ntgdi/bitmaps.c (hunk 4 above)
Comment (tkreuzer) [reply win32ss/gdi/ntgdi/bitmaps.c:340]: Yeah, that looks like copypasta. And it's there since 2014.

**Pull request merged**